keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1179 @@
+component reads a CEL expression from storage written by a (N+1) newer version
+(either due to version skew or a rollback), that a compatible CEL environment
+can still be constructed and the expression can still be evaluated.  This is
+achieved by making any CEL environment changes (language settings, libraries,
+variables) available for [stored expressions one version before they are allowed
+to be written by new
+expressions](https://github.com/kubernetes/kubernetes/blob/7fe31be11fbe9b44af262d5f5cffb1e73648aa96/staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go#L38).
+
+The only change we must make for this enhancement is to remove the
+[compatibility version
+constant](https://github.com/kubernetes/kubernetes/blob/7fe31be11fbe9b44af262d5f5cffb1e73648aa96/staging/src/k8s.io/apiserver/pkg/cel/environment/base.go#L45)
+and instead always pass in N-1 of the compatibility version introduced by this
+enhancement as the CEL compatibility version.
+
+### StorageVersion Compatibility Versioning
+
+StorageVersions specify what version an apiserver uses to write resources to etcd
+for each API group. The StorageVersion changes across releases as API groups
+graduate through stability levels.
+
+During upgrades and downgrades, the storage version is particularly important.
+To enable upgrades and rollbacks, the version selected for storage in etcd in 
+version N must be (en/de)codable for k8s versions N-1 through N+1.
+
+Thus, to determine the storage version to use at compatibility version N, we 
+will find the set of all supported GVRs for each of N-1, N, and N+1 and intersect 
+them to find a list of all GVRs supported by every binary version in the window. 
+The storage version of each group-resource is the newest 
+(using kube-aware version sorting) version found in that list for that group-resource.
+
+### API Compatibility Versioning
+
+Similar to feature flags, all APIs group-versions declarations will be modified
+to track which Kubernetes version the API group-versions are introduced (or
+removed) at.
+
+GA APIs should match the exact set of APIs enabled in GA for the Kubernetes version
+the compatibility version is set to.
+
+All Beta APIs (both off-by-default and on-by-default, if any of those
+still exist) need to behave exactly as they did for the Kubernetes version
+the compatibility version is set to. I.e. `--runtime-config=api/<version>` needs
+to be able to turn on APIs exactly like it did for each Kubernetes version that
+compatibility version can be set to.
+
+Alpha APIs may not be enabled in conjunction with compatibility version.
+
+### Discovery
+
+Discovery will [enable](https://github.com/kubernetes/kubernetes/blob/7080b51ee92f67623757534f3462d8ae862ef6fe/staging/src/k8s.io/apiserver/pkg/util/openapi/enablement.go#L32) the group versions matching the compatibility version.
+
+API fields that were introduced after the compatibility version will **not** be
+pruned. There is a tradoff here:
```

Comment:
There was some recent discussion in slack about at least annotating deprecation/feature availability for specific fields. Should this line be taken to mean that would be out of scope for this KEP (and require a second one)?

Reply:
Yeah, I think declarative field stability level tracking would be a great complement to this KEP, but it adds a lot of scope. I would like to see it happen at some point.